### PR TITLE
Document what's considered as a scalar type

### PR DESCRIPTION
**Pull request by ekinhbayar**
Initial take for issue #3939.

It made me go through all of the types documented on the type-system page, and this may be more than what the issue originally asks for. In which case, I can do a simpler attempt. Parts of this may be confusing or simply wrong too, if so, please correct me or let me know. This was a sort of a recap of theory and a categorization exercise which I may have failed : )

### CODE REVIEW

File: language/types/type-system.xml [hunk 1]
```diff
@@ -21,139 +21,192 @@
...
     </simpara>
+    <itemizedlist>
+     <listitem>
+      <simpara><type>false</type></simpara>
+     </listitem>
+     <listitem>
+      <simpara><type>true</type></simpara>
+     </listitem>
+    </itemizedlist>
    </listitem>
```
Comment (Girgias) [inline language/types/type-system.xml:88]: I'm back to wondering if we should call those `singleton` types, but that is orthogonal to this change.

File: language/types/type-system.xml [hunk 2]
```diff
@@ -21,139 +21,192 @@
...
+   <listitem>
+    <simpara><type>callable</type> type</simpara>
    </listitem>
   </itemizedlist>
 
-  <sect3 xml:id="language.types.type-system.composite.intersection">
-   <title>Intersection types</title>
+  <sect3 xml:id="language.types.type-system.atomic.scalar">
+   <title>Scalar types</title>
    <para>
```
Comment (Girgias) [inline language/types/type-system.xml:112]: Nit: to follow the new style guide
```suggestion
   <simpara>
```

File: language/types/type-system.xml [hunk 3]
```diff
@@ -21,139 +21,192 @@
...
-    <literal>V</literal> will be written as <literal>T&amp;U&amp;V</literal>.
+    A value is considered scalar if it is of type <type>int</type>,
+    <type>float</type>, <type>string</type> or <type>bool</type>.
+   </para>
+   <para>
+    While <literal>null</literal> value can only be of type <type>null</type>
+    and therefore is scalar, built-in function <function>is_scalar</function>
+    does not consider values of type <type>null</type> to be scalar.
+    To test for <literal>null</literal>, use the <function>is_null</function> function.
    </para>
```
Comment (Girgias) [inline language/types/type-system.xml:121]: I'm not sure I would mention `null` here, this might still make more sense as a caution on the `is_scalar()` docs.

File: language/types/type-system.xml [hunk 4]
```diff
@@ -21,139 +21,192 @@
...
-    <literal>V</literal> will be written as <literal>T|U|V</literal>.
-    If one of the types is an intersection type, it needs to be bracketed
-    with parenthesis for it to written in <acronym>DNF</acronym>:
-    <literal>T|(X&amp;Y)</literal>.
+    Types <type>never</type> and <type>void</type> can be declared as the
+    return type of a function to indicate whether the function does not
+    terminate or does not return a value. Their usage as a parameter
+    or a property type is not permitted.
    </para>
   </sect3>
```
Comment (Girgias) [inline language/types/type-system.xml:133]: Does this section really make that much sense? This is described on the respective pages already.

File: language/types/type-system.xml [hunk 5]
```diff
@@ -21,139 +21,192 @@
...
+      <simpara>Return-only types</simpara>
+      <itemizedlist>
+       <listitem>
+        <simpara><type>never</type> type</simpara>
+       </listitem>
+       <listitem>
+        <simpara><type>void</type> type</simpara>
+       </listitem>
+      </itemizedlist>
+     </listitem>
```
Comment (Girgias) [inline language/types/type-system.xml:67]: I don't know if it is that useful to indicate that those types are return only in the list of types?

File: language/types/type-system.xml (hunk 1 above)
Comment (ekinhbayar) [reply language/types/type-system.xml:88]: Just to fully confirm, is it on purpose that these "value types" are considered outside of the built-in types?

File: language/types/type-system.xml [hunk 6]
```diff
@@ -2,170 +2,201 @@
...
+ </simpara>
 
  <sect2 xml:id="language.types.type-system.atomic">
   <title>Atomic types</title>
-  <para>
+  <simpara>
    Some atomic types are built-in types which are tightly integrated with the
    language and cannot be reproduced with user defined types.
-  </para>
+  </simpara>
```
Comment (Girgias) [inline language/types/type-system.xml:22]: I meant to only use simpara for new text, to prevent work and large diff for translations, yes this is no ideal. Sorry for being unclear, if you could partially revert the change.

File: language/types/type-system.xml [hunk 7]
```diff
@@ -2,170 +2,201 @@
...
+   <title>User-defined types</title>
+   <simpara>
+    It is possible to define custom types with
+    <link linkend="language.oop5.interfaces">interfaces</link>,
+    <link linkend="language.oop5.basic.class">classes</link> and
+    <link linkend="language.types.enumerations">enumerations</link>.
+    These are considered as user-defined types, or class-types.
+    For example, a class called <literal>Elephant</literal> can be defined, then objects
+    of type <literal>Elephant</literal> can be instantiated, and a function can request
+    a parameter of type <literal>Elephant</literal>.
```
Comment (Girgias) [inline language/types/type-system.xml:123]: Semantic line breaks
```suggestion
    For example, a class called <literal>Elephant</literal> can be defined,
    then objects of type <literal>Elephant</literal> can be instantiated,
    and a function can request a parameter of type <literal>Elephant</literal>.
```

File: language/types/type-system.xml [hunk 8]
```diff
@@ -2,170 +2,201 @@
...
+   Values of type <type>array</type>, and <type>object</type> can be composed of multiple
+   atomic and composite elements.
+  </simpara>
+
+  <note>
+   <simpara>
+    The <type>resource</type> type is an opaque type, meaning it's internal structure
+    depends on the implementation. Therefore, it is considered neither scalar nor composite.
+   </simpara>
+  </note>
```
Comment (Girgias) [inline language/types/type-system.xml:141]: A resource is "like an object, just worse" so not exactly sure why it is mentioned here.

File: language/types/type-system.xml [hunk 9]
```diff
@@ -2,170 +2,201 @@
...
+   </simpara>
+  </sect3>
+ </sect2>
+
+ <sect2 xml:id="language.types.type-system.composite">
+  <title>Composite types</title>
+  <simpara>
+   It is possible to combine multiple atomic types into composite types.
+   Values of type <type>array</type>, and <type>object</type> can be composed of multiple
+   atomic and composite elements.
```
Comment (Girgias) [inline language/types/type-system.xml:133]: How can an array be composed of multiple atomic and composite types? It seems you are mixing elements/properties of array/objects with types which is incorrect.

A class is a type, an "array-shape" would be a type, what is inside an array/object is "irrelevant" from the type system PoV, and this is not what I was going for when writing this section about composite types, which is about how to combine different types together (e.g. `T|U`, `X&Y` and a type in DNF `array|(Traversable&Countable)`).

File: language/types/type-system.xml [hunk 10]
```diff
@@ -2,170 +2,201 @@
...
+    The <type>resource</type> type is an opaque type, meaning it's internal structure
+    depends on the implementation. Therefore, it is considered neither scalar nor composite.
+   </simpara>
+  </note>
+
+  <sect3 xml:id="language.types.type-system.composite.type-combination">
+   <title>Combining types</title>
+   <simpara>
+    PHP allows types to be combined in the following ways:
+   </simpara>
```
Comment (Girgias) [inline language/types/type-system.xml:147]: I don't understand why you are adding a sect3 here?

**Review by Girgias — COMMENTED** [review]
The diff GitHub is providing is... kinda garbage tried my best to comment on the content.

File: language/types/type-system.xml (hunk 1 above)
Comment (Girgias) [reply language/types/type-system.xml]: They are built-in, but they also represent a single *value* of a larger type.

For example, if we had singleton types for integers and a function only accepts a range of integers, we could write it like `function foo(1|2|3|4 $range1to4inclusive)`.

I did put them outside because they felt weird to be with the other built-in types, but they could be moved with them.

File: language/types/type-system.xml (hunk 6 above)
Comment (ekinhbayar) [reply language/types/type-system.xml:22]: I should've thought about translations, will indeed partially revert this, no worries.

File: language/types/type-system.xml (hunk 9 above)
Comment (ekinhbayar) [reply language/types/type-system.xml:133]: I am indeed mistaken on couple things here, one being element/property types vs containing more than its value, this is partly because I was reading https://phplang.org/spec/05-types.html as well. It is clear now though, knowing what's referred to as composite is really about unions and intersections, thank you. I'll revert changes to this section entirely.

File: language/types/type-system.xml (hunk 8 above)
Comment (ekinhbayar) [reply language/types/type-system.xml:141]: I think I was initially coming from the perspective of... tell the type of each listed item here so it's clear what's non-scalar, but it's definitely not necessary.

File: language/types/type-system.xml (hunk 1 above)
Comment (ekinhbayar) [reply language/types/type-system.xml]: I think I've addressed all but this one now. I could leave it as it is, though it may be nice to define it perhaps as simple as you just did : )

File: language/types/type-system.xml [hunk 11]
```diff
@@ -21,91 +21,108 @@
...
+     </listitem>
+     <listitem>
+      <simpara><link linkend="language.types.enumerations">Enumerations</link></simpara>
+     </listitem>
+    </itemizedlist>
+   </listitem>
+   <listitem>
+    <simpara><type>callable</type> type</simpara>
+   </listitem>
+  </itemizedlist>
```
[the remark this answers is not in the document]
Comment (ekinhbayar) [reply language/types/type-system.xml]: It sure did, I think the diff looks a lot better now.

**Comment by ekinhbayar** [comment]
hopefully the last ping for you on this @Girgias : ) I was wondering whether the new style guides documented somewhere, the contributing docs link on readme [can't be found](https://github.com/php/doc-base/blob/master/CONTRIBUTING_DOCS.md).

Also, at this stage of the PR, should I be requesting a review again?

**Comment by Girgias** [comment]
> hopefully the last ping for you on this @Girgias : ) I was wondering whether the new style guides documented somewhere, the contributing docs link on readme [can't be found](https://github.com/php/doc-base/blob/master/CONTRIBUTING_DOCS.md).
> 
> Also, at this stage of the PR, should I be requesting a review again?

Style guide is on the [doc.php.net](https://doc.php.net/guide/style.md) and has always been, but might be good to cross-link better.

No need to ask for rereviews as I get the email, just do it if I forget for a while. :)

**Comment by ekinhbayar** [comment]
Thanks for your time!
